Create a tutorial for custodians to use pooled accounts (#379)
stellar/new-docs: content/docs/building-apps/setup-custodial-account.mdx

---
title: Set Up a Custodial Account
order: 45
---

import { CodeExample } from "components/CodeExample";
import { Alert } from "components/Alert";

This guide describes how to add assets from the Stellar network to your custodial service. First, we walk through adding Stellar's native asset: lumens. Following that, we describe how to add other assets. This example uses Node.js and the [JS Stellar SDK](https://github.com/stellar/js-stellar-sdk), but it should be easy to adapt to other languages.

## Account Setup

### Pooled account
Most custodial services, including cryptocurrency exchanges, choose to use a single pooled Stellar account to handle transactions on behalf of their users instead of creating a new Stellar account for each customer. Generally, they keep track of their customers in a separate, internal database and use the memo field of a Stellar transaction to map an incoming payment to the corresponding internal customer.

> The benefit of using a pooled account is *lower costs* – no base reserves needed for each account – and *lower complexity* – you only need to manage one account key. However, with a pooled account it is your responsibility to manage all individual customer balances and payments. You can no longer rely on the Stellar ledger to accumulate value, handle errors and atomicity, and manage transactions on an account-by-account basis.

Reviewer: ```suggestion
The benefit of using a pooled account are *lower costs* – no base reserves are needed for each account – and *lower complexity* – you only need to manage one account key. However, with a pooled account, it is your responsibility to manage all individual customer balances and payments. You can no longer rely on the Stellar ledger to accumulate value, handle errors and atomicity, or manage transactions on an account-by-account basis.
```

 *     representation of a customer that's part of a pooled account but not
 *     on the Stellar network.
 * @param  {Asset}  asset   The asset being credited from the customer.
 * @param  {string} amount  The quantity of `asset` to withdraw.
 **/
function withdrawFromPool(from, asset, amount) {}

/**
 * Creates a PaymentOp corresponding to a withdrawal.
 *
 * This is just a convenience method for submitting payments on behalf of
 * customers in a transaction. You may want to just do one operation per
 * transaction (for example to include the source customer's ID in the memo),
 * or you can batch many operations together; it's up to you.
 *
 * @param  {string} to      The Stellar account address of the recipient
 * @param  {Asset}  asset   The asset to credit from the source customer
 * @param  {string} amount  The amount of `asset` to credit
 *
 * @return {xdr.paymentOp}
 */
function createPayment(to, asset, amount) {
  return sdk.Operation.payment({
    source: custodialAcc.accountId(),
    destination: to,
    asset: asset,
    amount: amount,
  });
}
```

</CodeExample>

## Integration points for listing XLM

### Setting the memo required flag on your account

> Before we get into the main integration points for adding XLM, make sure to configure your pooled account to require memos. This step is necessary to prevent users from forgetting to attach a memo, which can result in additional support needed from the custodial services and a less desirable user experience for new customers. [This article](https://www.stellar.org/developers-blog/fixing-memo-less-payments) explains how to set that up.

Reviewer: ```suggestion
Before we get into the main integration points for adding XLM support to your product or service, make sure to configure your pooled account to require memos. This step is necessary to prevent users from forgetting to attach a memo, which can increase customer support requests and a lead to a less desirable user experience for new customers. [This article](https://www.stellar.org/developers-blog/fixing-memo-less-payments) explains how to set that up.
```

### Listening for incoming payments from the Stellar network

> First, you need to listen for payments to the pooled account and debit any user that receives XLM there. For every payment received by the pooled account:

Reviewer: ```suggestion
First, you need to listen for payments to the pooled account and credit any user that receives XLM there. For every payment received by the pooled account:
```

  * check the memo field to determine which user should receive the payment, and

>   * debit the user’s account with the amount of XLM they received.

Reviewer: ```suggestion
  * credit the user’s account with the amount of XLM they received.
```

        const customerId = payment.transaction.memo;
        const amount = payment.amount;

        let asset = sdk.Asset.native();
        if (payment.asset_issuer) {
          asset = new sdk.Asset(payment.asset_code, payment.asset_issuer);
        }

        depositIntoPool(from, customerId, asset, amount);
      }
    });
```

</CodeExample>

When someone **outside** of your customer base wants to send lumens to one of your customers, instruct them to make an XLM payment to your pooled account address with the customer ID in the memo field of the transaction. Assuming you have set the `memo_required` configuration on your account (see [above](#setting-the-memo-required-flag-on-your-account)), [well-behaved](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0029.md) wallets should enforce it and prevent users from forgetting to attach memos to incoming payments. To be on the safe side, however, you should make it incredibly clear to senders that their payment will end up in limbo if they fail to attach a valid one. You can learn more about the transaction memo field [here](https://developers.stellar.org/docs/glossary/transactions/#memo).

> When someone **inside** of your customer base wants to send lumens to *another* one of your customers, you have two choices: you can send a memo'd payment exactly as above--this lets you maintain an audit trail, ensures the balance exists in the custodial account, etc.--or you can do the exchange "off-chain", i.e. by exclusively adjusting balances within your database.

Reviewer: ```suggestion
When someone **inside** of your customer base wants to send lumens to *another* one of your customers, you have two choices: you can send a memo'd payment exactly as above — this lets you maintain an audit trail, ensures the balance exists in the custodial account, etc. — or you can do the exchange "off-chain", i.e. by exclusively adjusting balances within your database.
```

 * Stellar address for the payment.
 */
function onPooledPayment(customerId, destination, amount, asset) {
  let tx = new sdk.TransactionBuilder(custodialAcc, {
    fee: 100,
    networkPassphrase: sdk.Networks.TESTNET,
    memo: customerId,
  })
    .addOperation(createPayment(destination, amount, asset))
    .setTimeout(30)
    .build();

  tx.sign(CUSTODIAL_KEY);
  return server.submitTransaction(tx).catch(onError).then((resp) => {
    return withdrawFromPool(customerId, asset, amount);
  });
}
```

</CodeExample>

This code would be called whenever one of your customers submitted a payment through your platform.

## Listing Other Stellar Assets
All of the code above is asset-agnostic, so accepting other (non-native/lumen) assets into your custodial account can be achieved after fulfilling a few prerequisities.

First, you must [open a trustline](https://developers.stellar.org/docs/start/list-of-operations/#change-trust) with the issuing account of the asset you’d like to list – without this, you cannot begin accepting the asset. The [Issue an Asset](https://developers.stellar.org/docs/issuing-assets/how-to-issue-an-asset/) tutorial covers the code for this one-time process.

> Next, if the asset issuer has the `authorization_required` flag on their account set, you will need to wait for them to authorized the trustline before you can begin accepting the asset. Read more about trustline authorization [here](https://developers.stellar.org/docs/issuing-assets/control-asset-access/).

Reviewer: ```suggestion
Next, if the asset issuer has the `authorization_required` flag on their account set, you will need to wait for them to authorize the trustline before you can begin accepting the asset. Read more about trustline authorization [here](https://developers.stellar.org/docs/issuing-assets/control-asset-access/).
```

<Alert>

> *Note*: Users cannot send a non-native asset to your pooled account (nor receive it from your customers) unless they have opted into that asset by opening a trustline to its issuer.

Reviewer: ```suggestion
*Note*: Users cannot send a non-native asset to your pooled account (nor receive one from your customers) unless they have opted into that asset by opening a trustline to its issuer.
```
